### PR TITLE
[FEATURE] DatasetLoader for BERT Pretraining

### PR DESCRIPTION
**Pull request by szhengac**
## Description ##
Add DatasetLoader that can process a bunch of files on-the-fly. The DatasetLoader maintains two worker pool, one for fetching and processing files, and the other one for generating mini-batch. With this one, we can train BERT using horovod without hanging.

### Essentials ###
- [X] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [X] Changes are complete (i.e. I finished coding on this PR)
- [X] All changes have test coverage
- [X] Code is well-documented

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/gluonnlp/data/datasetloader.py [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+
+
+def _manager_register():
+    BaseManager.register('ProxyArrayDataset', ProxyArrayDataset)
+
+
+class DatasetLoader:
+    """Loads data from a list of datasets and returns mini-batches of data.
+
+    One dataset is loaded at a time.
```
Comment (leezu) [inline src/gluonnlp/data/datasetloader.py]: How does this differ from the DatasetStream and prefetching logic?

Comment (szhengac) [reply src/gluonnlp/data/datasetloader.py]: This one maintains two worker pools for loading & *preprocessing* files and generating mini-batches. Here, *preprocessing* means training samples are generated on-the-fly, and reloading the same file can generate different training samples. The preprocessing function needs to be provided by the user. One particular examples can be seen from BERT pretraining script. And, it does not use queue, instead, it uses shared object among processes. Basically, it implements original gluon.dataloader while allowing generating new samples on-the-fly.

**Comment by mli** [comment]
Job PR-1099/2 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1099/2/index.html

**Comment by mli** [comment]
Job PR-1099/3 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1099/3/index.html

**Comment by mli** [comment]
Job PR-1099/5 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1099/5/index.html

**Comment by szhengac** [comment]
I have disabled name-error for `data/__init__.py`. Otherwise, it keeps giving me weird name error.

**Comment by mli** [comment]
Job PR-1099/6 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1099/6/index.html

**Comment by mli** [comment]
Job PR-1099/7 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1099/7/index.html

**Comment by mli** [comment]
Job PR-1099/8 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1099/8/index.html

**Comment by mli** [comment]
Job PR-1099/9 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1099/9/index.html

**Comment by eric-haibin-lin** [comment]
Would it be useful to add it to mxnet, too?

File: src/gluonnlp/data/sampler.py [hunk 2]
```diff
@@ -515,8 +516,12 @@ class SplitSampler(Sampler):
     even_size: bool, default False
       If the number of samples is not even across all partitions, sample a few extra samples
       for the ones with fewer samples.
+    repeat: int, default 1
+      The number of times that items are repeated.
+    shuffle: bool, default False
+      Whether or not to shuffle the items.
     """
-    def __init__(self, length, num_parts=1, part_index=0, even_size=False):
+    def __init__(self, length, num_parts=1, part_index=0, even_size=False, repeat=1, shuffle=False):
```
Comment (eric-haibin-lin) [inline src/gluonnlp/data/sampler.py]: Should we default shuffle=True? Otherwise it's not backward compatible

File: src/gluonnlp/data/__init__.py [hunk 3]
```diff
@@ -38,11 +38,12 @@
...
+from .datasetloader import *
 
 
 __all__ = (['batchify'] + utils.__all__ + transforms.__all__ + sampler.__all__
            + dataset.__all__ + corpora.__all__ + sentiment.__all__
            + word_embedding_evaluation.__all__ + stream.__all__ + conll.__all__
            + translation.__all__ + registry.__all__ + question_answering.__all__
            + dataloader.__all__ + candidate_sampler.__all__ + intent_slot.__all__
-           + glue.__all__)  # pytype: disable=attribute-error
+           + glue.__all__ + datasetloader.__all__)  # pytype: disable=attribute-error
```
Comment (eric-haibin-lin) [inline src/gluonnlp/data/__init__.py]: Shall we also update docs/api/modules/data.rst 
I noticed that SplitSampler is also missing there... I forgot to add it

File: tests/unittest/train/test_datasetloader.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        np.save(os.path.join('tests', 'data', 'part_{}.npy'.format(i)), np.random.uniform(size=(100, 20)))
+    data = os.path.join('tests', 'data')
+    split_sampler = nlp.data.SplitSampler(num_files, num_parts=1, part_index=0, shuffle=False)
+
+    dataset_params = {'allow_pickle': True}
+    sampler_params = {'batch_size': 2}
+    X = np.load(os.path.join(data, 'part_{}.npy'.format(0)))
+    for i in range(1, num_files):
+        X = np.concatenate((X, np.load(os.path.join(data, 'part_{}.npy'.format(i)))))
+    for num_dataset_workers in [1]:
```
Comment (eric-haibin-lin) [inline tests/unittest/train/test_datasetloader.py]: Shall we test other cases like num_worker=2 ? cache=True?

File: scripts/bert/pretraining_utils.py [hunk 5]
```diff
@@ -92,52 +90,59 @@ def get_model_loss(ctx, model, pretrained, dataset_name, vocab, dtype,
     model = BERTForPretrain(model, nsp_loss, mlm_loss, len(vocabulary))
     return model, vocabulary
 
-class BERTPretrainDataset(mx.gluon.data.ArrayDataset):
-    """Dataset for BERT pre-training.
 
-    Each record contains the following numpy ndarrays: input_ids, masked_lm_ids,
-    masked_lm_positions, masked_lm_weights, next_sentence_labels, segment_ids, valid_lengths.
+def prepare_pretrain_npy_dataset(filename, allow_pickle=False):
+    """Create dataset based on the files"""
```
Comment (eric-haibin-lin) [inline scripts/bert/pretraining_utils.py]: doc: Create dataset based on the npz files

File: scripts/bert/pretraining_utils.py [hunk 6]
```diff
@@ -92,52 +90,59 @@ def get_model_loss(ctx, model, pretrained, dataset_name, vocab, dtype,
...
+    assert not isinstance(filename, (list, tuple)), \
+        'When .npy/.npz data file is loaded, filename must be a string.'
+    logging.debug('start to load files %s ...', filename)
+    return nlp.data.NumpyDataset(filename, allow_pickle=allow_pickle)
+
+
+def prepare_pretrain_text_dataset(filename, tokenizer, max_seq_length, short_seq_prob,
+                                  masked_lm_prob, max_predictions_per_seq, whole_word_mask,
+                                  vocab, num_workers=1, worker_pool=None):
+    """Create dataset based on the files"""
```
Comment (eric-haibin-lin) [inline scripts/bert/pretraining_utils.py]: doc: Create dataset based on the raw textz files

File: src/gluonnlp/data/datasetloader.py [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+        Note that using large prefetching batch will provide smoother bootstrapping performance,
+        but will consume more shared_memory. Using smaller number may forfeit the purpose of using
+        multiple worker processes, try reduce `num_batch_workers` in this case.
+        By default it defaults to `num_batch_workers * 2`.
+    dataset_cached : bool, default is False
+        Whether or not to cache last processed dataset. Each processed dataset can
+        only be cached for once. When there is no new available processed dataset to be fetched,
+        we pop a cached processed dataset.
+    num_max_dataset_cached : int, default is 0
+        Maximum number of cached datasets. It is valid only if dataset_cached is True
```
Comment (eric-haibin-lin) [inline src/gluonnlp/data/datasetloader.py]: This class is fairly complicated for a new user to understand. Do we plan to have example in this doc, or adding it to the `api_notes/data.rst` to help users understand how to use this for large scale dataset?

**Comment by szhengac** [comment]
> Would it be useful to add it to mxnet, too?

I am considering adding it to mxnet in the future after the MxNet 2.0 dataloader is finished. What do you think? 

ping @zhreshold

File: src/gluonnlp/data/datasetloader.py (hunk 7 above)
Comment (szhengac) [reply src/gluonnlp/data/datasetloader.py]: Of course, I think we can use bert training as an example.

File: scripts/bert/run_pretraining.py [hunk 8]
```diff
@@ -128,8 +128,18 @@
 parser.add_argument('--sentencepiece', default=None, type=str,
                     help='Path to the sentencepiece .model file for both tokenization and vocab. '
                          'Effective only if --raw is set.')
-parser.add_argument('--num_data_workers', type=int, default=8,
-                    help='Number of workers to pre-process data.')
+parser.add_argument('--num_dataset_workers', type=int, default=4,
+                    help='Number of workers to pre-process dataset.')
+parser.add_argument('--num_batch_workers', type=int, default=4,
+                    help='Number of workers to pre-process mini-batch.')
+parser.add_argument('--circle_length', type=int, default=32,
```
Comment (eric-haibin-lin) [inline scripts/bert/run_pretraining.py]: this is too large.. let's use a smaller value

File: src/gluonnlp/data/datasetloader.py [hunk 9]
```diff
@@ -0,0 +1,424 @@
...
+                    data = zip(*data)
+                    return [default_batchify_fn(i) for i in data]
+                else:
+                    data = np.asarray(data)
+                    return nd.array(data, dtype=data.dtype)
+    num_dataset_workers : int
+        Number of worker process for dataset creation.
+    num_batch_workers : int
+        Number of worker process for batch creation.
+    pin_memory : boolean, default False
```
Comment (eric-haibin-lin) [inline src/gluonnlp/data/datasetloader.py]: we might want to mention this leads to more GPU memory consumption

**Comment by mli** [comment]
Job PR-1099/11 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1099/11/index.html

**Comment by mli** [comment]
Job PR-1099/12 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1099/12/index.html

**Comment by mli** [comment]
Job PR-1099/13 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1099/13/index.html

**Comment by eric-haibin-lin** [comment]
I got this error when I use this PR to pre-train from scratch:
```
[1,1]<stderr>:Traceback (most recent call last):
[1,1]<stderr>:  File "run_pretraining.py", line 457, in <module>
[1,1]<stderr>:    train(data_train, data_eval, model)
[1,1]<stderr>:  File "run_pretraining.py", line 340, in train
[1,1]<stderr>:    next_data_batch = next(data_train_iter)
[1,1]<stderr>:  File "/home/ubuntu/src/ads/src/gluonnlp/data/datasetloader.py", line 126, in __next__
[1,1]<stderr>:    batch = pickle.loads(ret.get())
[1,1]<stderr>:  File "/home/ubuntu/src/mxnet/python/mxnet/gluon/data/dataloader.py", line 61, in rebuild_ndarray
[1,1]<stderr>:    fd = fd.detach()
[1,1]<stderr>:  File "/usr/lib/python3.5/multiprocessing/resource_sharer.py", line 58, in detach
[1,1]<stderr>:    return reduction.recv_handle(conn)
[1,1]<stderr>:  File "/usr/lib/python3.5/multiprocessing/reduction.py", line 181, in recv_handle
[1,1]<stderr>:    return recvfds(s, 1)[0]
[1,1]<stderr>:  File "/usr/lib/python3.5/multiprocessing/reduction.py", line 154, in recvfds
[1,1]<stderr>:    raise EOFError
[1,1]<stderr>:EOFError
```

**Comment by leezu** [comment]
> I got this error when I use this PR to pre-train from scratch:
> 
> ```
> [1,1]<stderr>:Traceback (most recent call last):
> [1,1]<stderr>:  File "run_pretraining.py", line 457, in <module>
> [1,1]<stderr>:    train(data_train, data_eval, model)
> [1,1]<stderr>:  File "run_pretraining.py", line 340, in train
> [1,1]<stderr>:    next_data_batch = next(data_train_iter)
> [1,1]<stderr>:  File "/home/ubuntu/src/ads/src/gluonnlp/data/datasetloader.py", line 126, in __next__
> [1,1]<stderr>:    batch = pickle.loads(ret.get())
> [1,1]<stderr>:  File "/home/ubuntu/src/mxnet/python/mxnet/gluon/data/dataloader.py", line 61, in rebuild_ndarray
> [1,1]<stderr>:    fd = fd.detach()
> [1,1]<stderr>:  File "/usr/lib/python3.5/multiprocessing/resource_sharer.py", line 58, in detach
> [1,1]<stderr>:    return reduction.recv_handle(conn)
> [1,1]<stderr>:  File "/usr/lib/python3.5/multiprocessing/reduction.py", line 181, in recv_handle
> [1,1]<stderr>:    return recvfds(s, 1)[0]
> [1,1]<stderr>:  File "/usr/lib/python3.5/multiprocessing/reduction.py", line 154, in recvfds
> [1,1]<stderr>:    raise EOFError
> [1,1]<stderr>:EOFError
> ```

@szhengac would it make sense to add a test case that runs the script and would catch such issue?

**Comment by szhengac** [comment]
Haibin got this issue after running the script for more than 10 hours. I am
investigating this. This looks like that shared memory is run out. But I
didn’t observe it when I went through data.

**Comment by mli** [comment]
Job PR-1099/18 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1099/18/index.html

**Comment by mli** [comment]
Job PR-1099/19 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1099/19/index.html

**Comment by mli** [comment]
Job PR-1099/23 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1099/23/index.html

**Comment by mli** [comment]
Job PR-1099/25 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1099/25/index.html

**Comment by leezu** [comment]
Is the EOF issue fixed?

**Comment by szhengac** [comment]
> Is the EOF issue fixed?

The issue is caused by too many open files. But I didn't observe it when I run the training using python3.6. I will try python 3.5, which was used by haibin, and see if I can get the error.

**Comment by mli** [comment]
Job PR-1099/26 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1099/26/index.html

**Comment by leezu** [comment]
In that case we can simply advise users to increase the number of open file handles setting of their OS.

**Pull request merged**